Patch 2

Hexlord opened the pull request:
#### Description - What's this PR do?
Not sure whether square root of 2 is a setting, not sure if it corresponds to Real num type too, I would suggest moving both Pi and SquareRootTwo to Math.hpp

#### Impacts/Risks of These Changes?
Settings.hpp includes RealNum.hpp, thus no impact is forecasted.

louis-langholtz commented:
Thanks for feedback on this issue.

`SquareRootTwo` is currently used by `UnitVec2`. Admittedly I felt some displeasure about putting it into `Settings.hpp` but it works at least in terms of what includes what. Putting it into `RealNum.hpp` was actually my first choice but it's overridden by `RealNum.hpp.in`. I could put it in `RealNum.hpp.in` (and `RealNum.hpp`) but I really want to keep this file override mechanism more minimal than that (and keep it just for setting the `Real` type alias). My longer term preference is to reduce what's in `Settings.hpp` to solely being for things that are *settings* so my having put it in there was more of a hack just to stick it somewhere that at least worked. Moving `Pi` and `SquareRootTwo` into their own file called something like `RealConstants.hpp` is what I'm most keen on in terms of where to put them. `Math.hpp` I want to be for mathematical code that crosses multiple types.

I'm happy to hear more from anyone on this however.

What's particularly interesting for me is understanding people's feelings about their experiences with software. Like does it feel surprising and/or confusing to find `SquareRootTwo` in `Settings.hpp`? Does it generally feel straight forward where other things are? Does it feel easier to setup a basic simulation that was expected?

My preference - speaking generally - is for the experience of using the software to feel as straight forward, easy to use, and impressive as it can be made. The *funner* the *better*! :smiley:

louis-langholtz commented on PlayRho/Common/RealNum.hpp:
```diff
@@ -73,6 +73,10 @@ namespace playrho
+    /// @brief SquareRootTwo.
+    ///
+    constexpr auto SquareRootTwo = Real(1.414213562373095048801688724209698078569671875376948073176679737990732478462);
+
```
Unfortunately adding `SquareRootTwo` to this file won't work unless it's also added to the `RealNum.hpp.in` file. `*.in` files should be files that override their namesakes by `cmake` (or any other such facility). And loading override files with anything other than what's being set at build time, is less appealing to me than the constant being in the `Settings.hpp` file.

louis-langholtz commented:
Along this idea... I'm working on moving the memory stuff (Alloc, Realloc, Free) out from Settings and into its own files.

Hexlord commented:
> What's particularly interesting for me is understanding people's feelings about their experiences with software. Like does it feel surprising and/or confusing to find SquareRootTwo in Settings.hpp? Does it generally feel straight forward where other things are? Does it feel easier to setup a basic simulation that was expected?

Actually in my project I have included your Math.hpp in my Math.hpp and exposed the namespace,
same for my Units.hpp, but there I have also added Integer and Precise boost-ed types, so all this inner dances with variable positions does affect esthetics not usability. 

I have also added Vector4D and changed Vec234 to Real234, along with creating Integer234 and Precise234. This partially defeats the purpose of unit system compile time control (working only for Real), but yet it is still useful for common purposes. This extra staff should not be in project not actually using it, but I could share it if you'd like.

louis-langholtz commented:
@Hexlord I've just gone through some of the changes you've done in your fork. Some interesting work! Here's some reflections on what I saw:

**Mispellings:** Seems you found some misspellings and corrected them. If you make a branch just for spelling corrections and make pull requests for those, they'd be easy for me to review and very appealing to me to accept!

**Pi:** This one repeatedly takes me by surprise. It's actually a unit-less constant (the ratio of the circumference of a circle to its diameter) but almost every time I go back to it after not having looked at it in a while I think it should be expressed in terms of radians. Here's some links on this: [mathhelpforum](http://mathhelpforum.com/geometry/255678-what-unit-pi.html), [WikiPedia](https://en.wikipedia.org/wiki/Pi). So that's why I haven't expressed it in terms of any units in my code (not even in terms of the radian).

**`Vec[2,3]` vs. `Real[2,3]`**: I see you renamed the `Vec2` and `Vec3` structures to `Real2` and `Real3`. I left the names of `Vec2` and `Vec3` behind to tie into the history of having come from the `b2Vec2` and `b2Vec3` names. I'm not adamant on keeping the names as `Vec2` and `Vec3` however. Really the primary reason I want there to be any use at all anymore of `Vec2` or `Vec3` is to give users a backward compatible way of expressing unitless vectors so keeping the name closer to what it used to be is appealing that way. Generally speaking however, I'd prefer all values to be expressed in terms of their units which is why I made the types like `Length` and `Meter`. Unfortunately, by default, these unit types aren't compile time enforced since I've found that the current C++ compiler technology isn't actually able to always do zero-overhead w.r.t. the Boost Units library; i.e. at least with clang, there's a performance penalty for using boost units so I haven't enabled it by default.

Hexlord commented:
Will do.

Was fighting with `constexpr auto EarthlyGravity = LinearAcceleration2D{` for hours, now you mentioning performance effect, I give up with boost units, my MSVS really hates this trickery *_*. It was fun getting to discover that crazy stuff a bit though. 

The vector problem is really sharp, your attempt is scalable, yet not very usability-ful, having a lot of procedures accessing X or Y fields is hard in context of your vector interface, yet I have made little code to help with this, I will try to sumbit it in a different branch for you to take a look later.

Yes, PI really is a ratio value, thanks for that. 

Not really the topic for it, but here you hear my thoughts:

I would love to research your multi-threading attempts deeper. You mentioned, if I remember correctly, that it is no win against single threading inside PlayRho, and in some thread it was 2% and 7% for other's attempts, yet how about a situation where you have multiple Worlds. And I think Body transition between World's thus leading to contact with Chunking could open borders for simulations of big spaces a little wider.

Fluid is a thing to go, as already mentioned in some topic. The effective solution as I see it are circles charging close cells of water gravity field to simulate viscosity. The rest is about rendering shaders. Mathematical solutions seem to affect performance way too much, can not involve GPU into calculations neither.

Last, but not the least is Moving Stuff. Functionality should appear where after manually moving dynamic body it stops simulating for that very frame of teleportation, and the next it smoothly appears in the physical world (not to get stuck it must displace dynamic bodies from its insides, then displace itself from static bodies if any. Yes, a huge problem are bodies being stuck in eachother after crazyfast impacts. Some secondary mechanic besides common physics resolvers should take its place to resolve anomalies. Like remember the normal of contact just before the anomaly and only allow displacement in that direction instead of jagging crazzinness.
This is very common in games, yet in box2d engine teleportation lead to some crazy behaviours in terms of performance.

I would be happy to share my advancements if any, because transition from raw box2d to playrho seems like a big step towards much better performance (-Ofast and stuff being safer), much easier extansibility, developer that is breathing with answers on the net, which is extremely pleasant to observe!

louis-langholtz commented:
A `playrho::World` instance is currently fully independent from other instances. Meaning that there should be no problem what-so-ever with running an application with multiple world instances where each instance is in its own thread. Not sure if that's already well recognized.

While I don't believe this is 100% the case for Erin's Box2D `b2World` instances -- since they share globally writable data -- I'm also not aware that running multiple world instances is that interesting to users. If it is... wow, that'd be so cool to me since at least at the world instance granularity `playrho` should be 100% safe to run in separate threads.

Hexlord commented:
@louis-langholtz That is a sure thing, I am using like 10+ b2Worlds at a time, the problem is in transition between them. Removing body and adding body to another world raises stuck-problems and affects performance the known way, and requires to solve the edge transition problem. Something like dublicating bodies in shared world scenes (at edges) could be executed, but at immerse cost of implementation complexity. I mean just before leaving World1, one half of a box does already belong to World2 and must collide with its stuff...

Yet if you have 2 layers of gameplay phys objects of which should never collide between those layers you can affectively use 2 seperate Worlds to increase performance in a big way. I did not combine this with multithreading though, way to go. The complexity is actors having ability to gravitationally impact ALL layers, thus leading to whether a mutex hell or seperate stages of physical update, complexity raises yet again.

Also I remember one more idea: Heavy weight big bodies may ignore little weight small bodies ignoring collison resolving just displacing them out as a kinematic body would do, thus saving the day in a situation of large meteor flying through 100 glass pieces. That could be a huge performance boost for a frequent situation in my simulations, going to look towards that too.
